src/connection.rs

```diff
@@ -0,0 +1,462 @@
+
+    fn next_stream_id(&mut self) -> Result<stream::Id, ConnectionError> {
+        if self.id_counter >= u32::MAX as usize - 2 {
+            return Err(ConnectionError::NoMoreStreamIds)
+        }
+        let proposed = stream::Id::new(self.id_counter as u32);
+        self.id_counter += 2;
+        match self.mode {
+            Mode::Client => assert!(proposed.is_client()),
+            Mode::Server => assert!(proposed.is_server())
+        }
+        Ok(proposed)
+    }
+
+    fn is_valid_remote_id(&self, id: stream::Id, ty: Type) -> bool {
+        match ty {
+            Type::Ping | Type::GoAway => return id.is_session(),
+            _ => {}
+        }
+        match self.mode {
+            Mode::Client => id.is_server(),
+            Mode::Server => id.is_client()
+        }
+    }
+
+    fn new_stream(&mut self, id: stream::Id, recv_window: u32) -> Stream {
+        let recv_win = Arc::new(Window::new(AtomicUsize::new(recv_window as usize)));
+        let (stream_tx, stream_rx) = mpsc::unbounded();
+        let inbox = StreamHandle {
+            recv_win: recv_win.clone(),
+            sender: stream_tx,
+            ack: true
+        };
+        self.streams.insert(id, inbox);
+        Stream::new(id, self.config.clone(), self.stream_tx.clone(), stream_rx.fuse(), recv_win)
+    }
+
+    fn deliver(&mut self, id: stream::Id, item: Item) -> bool {
+        if let Some(ref inbox) = self.streams.get(&id) {
+            if inbox.sender.unbounded_send(item).is_ok() {
+                return true
+            }
+        }
+        trace!("can not deliver; stream {} is gone", id);
+        self.streams.remove(&id);
+        false
+    }
+
+
+    fn terminate(&mut self) {
+        debug!("terminating connection");
+        self.is_dead = true;
+        self.streams.clear()
+    }
+
+    fn send(&mut self, frame: RawFrame) -> Poll<(), ConnectionError> {
+        trace!("send: {:?}", frame);
+        match self.resource.start_send(frame) {
+            Ok(AsyncSink::Ready) => Ok(Async::Ready(())),
+            Ok(AsyncSink::NotReady(frame)) => {
+                self.pending = Some(frame);
+                Ok(Async::NotReady)
+            }
+            Err(e) => {
+                self.terminate();
+                Err(e.into())
+            }
+        }
+    }
+
+    fn flush(&mut self) -> Poll<(), ConnectionError> {
+        self.resource.poll_complete().map_err(|e| {
+            self.terminate();
+            e.into()
+        })
+    }
+}
+
+impl<T> futures::Stream for Connection<T>
+where
+    T: AsyncRead + AsyncWrite
+{
+    type Item = Stream;
+    type Error = ConnectionError;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        if self.is_dead {
+            return Ok(Async::Ready(None))
+        }
+
+        // First, check for pending frames we need to send.
+        if let Some(frame) = self.pending.take() {
+            trace!("send pending: {:?}", frame);
+            try_ready!(self.send(frame))
```

Comment:
If this returns `NotReady`, `self.pending` will be empty at the next iteration.

Reply:
Oh nevermind, it will set `self.pending` back to the same frame.

Reply:
No, because `Connection::send` puts its argument back into `Connection::pending` before returning `NotReady`.

Reply:
Sorry for the comment. GitHub preferred to not show me your second comment.